Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,1047 @@
+To do this we will make use of the `integers` Hypothesis strategy.
+Let's include a bad assertion statement – that `y` can't be larger than 25 – to see how Hypothesis reports this fail case.
+Note that we aren't really testing any code here, we are simply exercising some of the tools that Hypothesis provides us with.
+
+```python
+from hypothesis import given 
+import hypothesis.strategies as st
+
+# using `given` with multiple parameters
+# `x` is an integer drawn from [0, 10]
+# `y` is an integer drawn from [20, 30]
+@given(x=st.integers(0, 10), y=st.integers(20, 30))
+def test_demonstrating_the_given_decorator(x, y):
+    assert 0 <= x <= 10
+    
+    # `y` can be any value in [20, 30]
+    # this is a bad assertion: it should fail!
+    assert 20 <= y <= 25
+```
+
+See that the names of the parameters specified in `given` — `x` and `y` in this instance — must match those in the signature of the test function.
+
+To run this test function, we simply call `test_demonstrating_the_given_decorator()`.
+Note that, unlike with a typical function, we do not pass values for `x` and `y` to this function – *the* `given` *decorator will pass these values to the function for us*.
+Executing this `given`-decorated function will prompt Hypothesis to draw 100 pairs of values for `x` and `y`, according to their respective strategies, and the body of the test will be executed for each such pair.
+
+```python
+# running the test
+>>> test_demonstrating_the_given_decorator()
+Falsifying example: test_demonstrating_the_given_decorator(
+    x=0, y=26,
+)
+---------------------------------------------------------------------------
+AssertionError                            Traceback (most recent call last)
+<ipython-input-29-ea20353bbef5> in test_demonstrating_the_given_decorator(x, y)
+     10     # `y` can be any value in [20, 30]
+     11     # this is a bad assertion: it should fail!
+---> 12     assert 20 <= y <= 25
+
+AssertionError: 
+
+```
+
+(Note: some of the "Traceback" error message has been removed to improve legibility)
+
+See that the error message here indicates that Hypothesis identified a "falsifying example", or a set of input values, `x=0` and `y=26`, which caused our test function to raise an error. The proceeding "Traceback" message indicates that it is indeed the second assertion statement that is responsible for raising the error.
+
+### Shrinking: Simplifying Falsifying Inputs
+
+The `given` decorator strives to report the "simplest" set of input values that produce a given error.
+It does this through the process of "shrinking".
+Each of Hypothesis' strategies has its own prescribed shrinking behavior.
+For the `integers` strategy, this means identifying the integer closest to 0 that produces the error at hand.
+For instance, `x=12` and `y=29` may have been the first drawn values to trigger the assertion error.
```

Comment:
not really that relevant to the discussion right here but are you planning to talk at all about how hypothesis actually draws data?

Reply:
@Zac-HD do you think there is anything sufficiently concise and informative to say towards this end, given the basic nature of this discussion?

Reply:
Nope... I'd just leave it totally unspoken and they can go to Hypothesis documentation if curious.

Which is almost as silent on the topic, just "it's reproducible, arbitrary, and don't try to guess the distribution".  Anything further is internals and subject to change without notice.